File: libs/language/src/lib/merge/processors/ValuePropagator.ts
import { isNonNullable, isNullable } from '@blue-labs/shared-utils';
import { BlueNode } from '../../model';
import { MergingProcessor } from '../MergingProcessor';
import {
  isBigDecimalNumber,
  isBigIntegerNumber,
} from '../../../utils/typeGuards';

/**
 * Propagates values from source to target nodes and throws an error if there is a conflict
 */
export class ValuePropagator implements MergingProcessor {
  process(target: BlueNode, source: BlueNode): BlueNode {
    const sourceValue = source.getValue();

    if (isNonNullable(sourceValue)) {
      const targetValue = target.getValue();
      if (isNullable(targetValue)) {
        return target.cloneShallow().setValue(sourceValue);
      } else if (!isEqualValue(sourceValue, targetValue)) {
        if (hasOverrideableDefault(target)) {
          return target.cloneShallow().setValue(sourceValue);
        }
        throw new Error(
          `Node values conflict. Source node value: ${sourceValue}, target node value: ${targetValue}`,
        );
      }
    }
    return target;
  }
}

const isEqualValue = (a: BlueNode['value'], b: BlueNode['value']) => {
  if (
    (isBigIntegerNumber(a) && isBigIntegerNumber(b)) ||
    (isBigDecimalNumber(a) && isBigDecimalNumber(b))
  ) {
    return a.eq(b);
  }
  return a === b;
};

const hasOverrideableDefault = (node: BlueNode): boolean =>
  isNonNullable(node.getType()) ||
  isNonNullable(node.getItemType()) ||
  isNonNullable(node.getKeyType()) ||
  isNonNullable(node.getValueType()) ||
  isNonNullable(node.getSchema()) ||
  isNonNullable(node.getDescription());

File: references/java-sdk/src/main/java/blue/language/merge/processor/ValuePropagator.java
package blue.language.merge.processor;

import blue.language.NodeProvider;
import blue.language.model.Node;
import blue.language.merge.MergingProcessor;
import blue.language.merge.NodeResolver;

public class ValuePropagator implements MergingProcessor {
    @Override
    public void process(Node target, Node source, NodeProvider nodeProvider, NodeResolver nodeResolver) {
        if (source.getValue() != null) {
            if (target.getValue() == null)
                target.value(source.getValue());
            else if (!source.getValue().equals(target.getValue()))
                throw new IllegalArgumentException("Node values conflict. Source node value: " + source.getValue() +
                        ", target node value: " + target.getValue());
        }
        
    }
}
